### PR TITLE
8315373: Change VirtualThread to unmount after freezing, re-mount before thawing

### PR DESCRIPTION
**Pull request by shipilev**
This is a series of Loom backports that brings in the fixes from JDK 22 and mainline into JDK 21, where Loom is production-ready. This batch fixes a few thread state and JVMTI problems.

Unfortunately, these backports are tied together, as they change the same code a few times. It is very hard to make them into separate backports without breaking stuff along the way, so instead I clobbered them together in this PR. They follow the application order from mainline. All patches together touch only the Virtual Thread parts, so the risk for the rest of the code should be minimal.

Brief explanation of fixes:

* [JDK-8315373](https://bugs.openjdk.org/browse/JDK-8315373): In mainline since Sep 2023, there is no bug tail. Changes how VTs interact with JVMTI. We want this fix to provide the ground for the rest of backports. Notably, it makes [JDK-8312777](https://bugs.openjdk.org/browse/JDK-8312777) backport clean. 
* [JDK-8312498](https://bugs.openjdk.org/browse/JDK-8312498): In mainline since Sep 2023. There is a bug tail, [JDK-8322818](https://bugs.openjdk.org/browse/JDK-8322818). Fixes what thread state VT reports when timed-parked. This fixes the actual Loom bug.
* [JDK-8312777](https://bugs.openjdk.org/browse/JDK-8312777): In mainline since Oct 2023, there is no bug tail. Fixes the JVMTI race with mount/unmount events. Makes [JDK-8322818](https://bugs.openjdk.org/browse/JDK-8322818) backport clean. This fixes the actual Loom bug.
* [JDK-8321270](https://bugs.openjdk.org/browse/JDK-8321270): In mainline since Dec 2023, there is no bug tail. Makes [JDK-8322818](https://bugs.openjdk.org/browse/JDK-8322818) a clean backport. This fixes the actual Loom performance problem.
* [JDK-8322818](https://bugs.openjdk.org/browse/JDK-8322818): In mainline since Jan 2023. Fixes the bug introduced by [JDK-8312498](https://bugs.openjdk.org/browse/JDK-8312498). Has two additional follow-ups: [JDK-8323002](https://bugs.openjdk.org/browse/JDK-8323002) and [JDK-8323296](https://bugs.openjdk.org/browse/JDK-8323296).
* [JDK-8323002](https://bugs.openjdk.org/browse/JDK-8323002): In mainline since Jan 2023. Follow-up test fix for [JDK-8322818](https://bugs.openjdk.org/browse/JDK-8322818). Test-only change.
* [JDK-8323296](https://bugs.openjdk.org/browse/JDK-8323296): In mainline since Jan 2023. Follow-up test fix for [JDK-8322818](https://bugs.openjdk.org/browse/JDK-8322818). Test-only change.
* [JDK-8316924](https://bugs.openjdk.org/browse/JDK-8316924): In mainline since Sep 2023. Follow-up test fix for [JDK-8322818](https://bugs.openjdk.org/browse/JDK-8322818). Test-only change.

Before I undraft this, which would trigger JBS updates:

@AlanBateman, does this make sense to you?

@jerboaa, how do you feel about this?

Additional testing:
 - [x] MacOS AArch64 server fastdebug, `jdk_loom hotspot_loom`
 - [x] Linux AArch64 server fastdebug, `jdk_loom hotspot_loom`
 - [x] Linux x86_64 server fastdebug, `jdk_loom hotspot_loom`
 - [x] Linux x86_64 server fastdebug, `tier{1,2,3}`

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] [JDK-8315373](https://bugs.openjdk.org/browse/JDK-8315373) needs maintainer approval
- [x] [JDK-8322818](https://bugs.openjdk.org/browse/JDK-8322818) needs maintainer approval
- [x] [JDK-8323296](https://bugs.openjdk.org/browse/JDK-8323296) needs maintainer approval
- [x] [JDK-8312498](https://bugs.openjdk.org/browse/JDK-8312498) needs maintainer approval
- [x] [JDK-8321270](https://bugs.openjdk.org/browse/JDK-8321270) needs maintainer approval
- [x] [JDK-8312777](https://bugs.openjdk.org/browse/JDK-8312777) needs maintainer approval
- [x] [JDK-8316924](https://bugs.openjdk.org/browse/JDK-8316924) needs maintainer approval
- [x] [JDK-8323002](https://bugs.openjdk.org/browse/JDK-8323002) needs maintainer approval

### Issues
 * [JDK-8315373](https://bugs.openjdk.org/browse/JDK-8315373): Change VirtualThread to unmount after freezing, re-mount before thawing (**Enhancement** - P4 - Approved)
 * [JDK-8312498](https://bugs.openjdk.org/browse/JDK-8312498): Thread::getState and JVM TI GetThreadState should return TIMED_WAITING virtual thread is timed parked (**Bug** - P3 - Approved)
 * [JDK-8312777](https://bugs.openjdk.org/browse/JDK-8312777): notifyJvmtiMount before notifyJvmtiUnmount (**Bug** - P3 - Approved)
 * [JDK-8321270](https://bugs.openjdk.org/browse/JDK-8321270): Virtual Thread.yield consumes parking permit (**Bug** - P3 - Approved)
 * [JDK-8322818](https://bugs.openjdk.org/browse/JDK-8322818): Thread::getStackTrace can fail with InternalError if virtual thread is timed-parked when pinned (**Bug** - P3 - Approved)
 * [JDK-8323002](https://bugs.openjdk.org/browse/JDK-8323002): test/jdk/java/lang/Thread/virtual/stress/GetStackTraceALotWhenPinned.java times out on macosx-x64 (**Bug** - P4 - Approved)
 * [JDK-8323296](https://bugs.openjdk.org/browse/JDK-8323296): java/lang/Thread/virtual/stress/GetStackTraceALotWhenPinned.java#id1 timed out (**Bug** - P4 - Approved)
 * [JDK-8316924](https://bugs.openjdk.org/browse/JDK-8316924): java/lang/Thread/virtual/stress/ParkALot.java times out (**Bug** - P4 - Approved)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk21u-dev.git pull/245/head:pull/245` \
`$ git checkout pull/245`

Update a local copy of the PR: \
`$ git checkout pull/245` \
`$ git pull https://git.openjdk.org/jdk21u-dev.git pull/245/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 245`

View PR using the GUI difftool: \
`$ git pr show -t 245`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk21u-dev/pull/245.diff">https://git.openjdk.org/jdk21u-dev/pull/245.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk21u-dev/pull/245#issuecomment-1946021578)

### CODE REVIEW

**Comment by shipilev** [comment]
All these backports are clean.

/clean

**Comment by jerboaa** [comment]
> @jerboaa, how do you feel about this?

This is porting work, so it would be advisable to get reviews from loom devs on this before considering for approval (even if it's clean).

**Comment by AlanBateman** [comment]
> Assuming Alan is fine with it, do you see other problems with the PR like this?

Normally I think some of these changes would only be backport if really need but given the churn in this newish code then it is safer to backport these as a group.  I looked through the diffs and everything looks okay. You may have to add JDK-8316924 to the list as the test ParkALot had to be dialed down for larger test systems.

**Comment by shipilev** [comment]
> > Assuming Alan is fine with it, do you see other problems with the PR like this?
> 
> Normally I think some of these changes would only be backport if really need but given the churn in this newish code then it is safer to backport these as a group. I looked through the diffs and everything looks okay. 

Thank you!

@jerboaa, anything else you want to be covered before I start doing the formal fix-requests for 21u-dev?

> You may have to add JDK-8316924 to the list as the test ParkALot had to be dialed down for larger test systems.

Right. Added.

**Comment by jerboaa** [comment]
@shipilev Update: The general consensus to this seems to be that there are no objections. It would be preferable to get this into 21.0.4 (early after 21.0.3 rampdown starts in end of Feb) unless there are very good reasons to get it into 21.0.3.

**Comment by shipilev** [comment]
> Update: The general consensus to this seems to be that there are no objections. It would be preferable to get this into 21.0.4 (early after 21.0.3 rampdown starts in end of Feb) unless there are very good reasons to get it into 21.0.3.

Thanks for update!

I think there is a reason to get it to 21.0.3. The major driver for JDK 21 is Loom, which was made production-ready in JDK 21. The work done in 21u right now is effectively stabilizing Loom for production use. Shipping 21.0.3 with known bugs would likely delay wide-scale production testing/usage to 21.0.4. This is the crux of the problem, I think: 21u is what customers would ultimately test Loom with.

This stack of patches was in mainline for a while, and only touches the Loom paths, so the risk appears to be low for non-Loom usages. Yes, there is a minor risk of Loom regressions, but unless we assume that JDK 22 GA would be widely used/tested, we would not be in better position for these backports in 21.0.4.

**Comment by shipilev** [comment]
/approval request JDK-8315373 Changes how VTs interact with JVMTI. In mainline since Sep 2023, there is no bug tail. We want this fix to provide the ground for the rest of backports. Notably, it makes JDK-8312777 backport clean. The fix is part of large atomic change, see 21u PR.

/approval request JDK-8312498 Fixes what thread state VT reports when timed-parked. This fixes the actual Loom bug.
In mainline since Sep 2023. There is a bug tail, JDK-8322818. The fix is part of large atomic change, see 21u PR.

/approval request JDK-8312777 Fixes the JVMTI race with mount/unmount events. This fixes the actual Loom bug. In mainline since Oct 2023, there is no bug tail. Makes JDK-8322818 backport clean. The fix is part of large atomic change, see 21u PR.

/approval request JDK-8321270 This fixes the actual Loom performance problem. In mainline since Dec 2023, there is no bug tail. Makes JDK-8322818 a clean backport. The fix is part of large atomic change, see 21u PR.

/approval request JDK-8322818 Fixes the bug introduced by JDK-8312498. In mainline since Jan 2023. Has two additional follow-ups: JDK-8323002 and JDK-8323296. The fix is part of large atomic change, see 21u PR.

/approval request JDK-8323002 Follow-up test fix for JDK-8322818. In mainline since Jan 2023. Test-only change. The fix is part of large atomic change, see 21u PR.

/approval request JDK-8323296 Follow-up test fix for JDK-8322818. In mainline since Jan 2023. Test-only change. The fix is part of large atomic change, see 21u PR.

/approval request JDK-8316924 Follow-up test fix for JDK-8322818. In mainline since Sep 2023. Test-only change. The fix is part of large atomic change, see 21u PR.

**Comment by shipilev** [comment]
OK, I think I misread how approval command works. Let me try again.

/approval JDK-8315373 request Changes how VTs interact with JVMTI. In mainline since Sep 2023, there is no bug tail. We want this fix to provide the ground for the rest of backports. Notably, it makes JDK-8312777 backport clean. The fix is part of large atomic change, see 21u PR.

/approval JDK-8312498 request Fixes what thread state VT reports when timed-parked. This fixes the actual Loom bug.
In mainline since Sep 2023. There is a bug tail, JDK-8322818. The fix is part of large atomic change, see 21u PR.

/approval JDK-8312777 request Fixes the JVMTI race with mount/unmount events. This fixes the actual Loom bug. In mainline since Oct 2023, there is no bug tail. Makes JDK-8322818 backport clean. The fix is part of large atomic change, see 21u PR.

/approval JDK-8321270 request This fixes the actual Loom performance problem. In mainline since Dec 2023, there is no bug tail. Makes JDK-8322818 a clean backport. The fix is part of large atomic change, see 21u PR.

/approval JDK-8322818 request Fixes the bug introduced by JDK-8312498. In mainline since Jan 2023. Has two additional follow-ups: JDK-8323002 and JDK-8323296. The fix is part of large atomic change, see 21u PR.

/approval JDK-8323002 request Follow-up test fix for JDK-8322818. In mainline since Jan 2023. Test-only change. The fix is part of large atomic change, see 21u PR.

/approval JDK-8323296 request Follow-up test fix for JDK-8322818. In mainline since Jan 2023. Test-only change. The fix is part of large atomic change, see 21u PR.

/approval JDK-8316924 request Follow-up test fix for JDK-8322818. In mainline since Sep 2023. Test-only change. The fix is part of large atomic change, see 21u PR.

**Comment by jerboaa** [comment]
Yes, integrating it sooner rather than later would be preferred. Ramp down starts next week for JDK 21.0.3. It's not a low risk change. Having said that, I concur it's fairly well isolated to loom code and there were no objections in principle from other maintainers. We'll have to see and if worse comes to worst we'll back it out before April.

So in a nutshell, the question was whether to allow for 21.0.3 or push to 21.0.4. It'll have about 2 months before the release in April (21.0.3). If it was pushed to 21.0.4 we'd have some more testing in the `jdk21u-dev` tree, but to me the first merge to `jdk21u` matters. Going by that, we'd have similarly about 2 months of testing with public EA builds. So the characteristics don't change much. Therefore, approved for `21.0.3`.

**Comment by gnu-andrew** [comment]
> So in a nutshell, the question was whether to allow for 21.0.3 or push to 21.0.4. It'll have about 2 months before the release in April (21.0.3). 

It's five weeks until code freeze for 21.0.3 and so five weeks to both find any issues and get any follow-on fixes in. The release happens on 2024-04-16 so in two months time it will be out.

> If it was pushed to 21.0.4 we'd have some more testing in the `jdk21u-dev` tree, but to me the first merge to `jdk21u`
> matters. Going by that, we'd have similarly about 2 months of testing with public EA builds. So the characteristics don't change much. Therefore, approved for `21.0.3`.

That would be a full two months. Even going from the initial build promotion, you've lost three weeks of development with 21.0.3 compared with 21.0.4 (the fourth build promotion is due to happen tomorrow).

If you want this in 21.0.3, I think it's imperative it gets merged as soon as possible to maximise the time available, ideally today.  Personally, I would still prefer 21.0.4.

**Comment by shipilev** [comment]
After discussing this with other engineers, and seeing more of the soft push back here, I am now leaning to keep this PR out of 21.0.3 and push it early into 21.0.4, some time next week. This would be less stressful for everyone involved.

The reason to have these patches is to make 21u conveniently usable for Loom adopters. However, there are other fixes (e.g. JDK-8320275) that would not make it to 21.0.3. Plus, there are dependent fixes (e.g. JDK-8322846) that we also want to get in after this lands. So backporting the performance fixes when there are not yet backported functional fixes does not gain us much ground for Loom stabilization for 21.0.3.

**Comment by shipilev** [comment]
Heads-up: I remerged from 21.0.4 master, re-testing with `all` tests now, will push once tests are confirmed green.
